File: oblig_8_GAME_OF_LIFE/rutenett.py
from random import random # Her bestemte jeg meg om å bruke random.
from celle import Celle

class Rutenett:
    def __init__(self, ant_rader, ant_kolonner):
        self._ant_rader = ant_rader
        self._ant_kolonner = ant_kolonner
        self._rutenett = self._lag_tomt_rutenett()

    def _lag_tom_rad(self):
        return [None for _ in range(self._ant_kolonner)]

    def _lag_tomt_rutenett(self):
        return [self._lag_tom_rad() for _ in range(self._ant_rader)]
    
    def lag_celle(self):
        celle = Celle()
        if random() <= 0.33:
            celle.sett_levende()
        return celle
    
    def fyll_med_tilfeldige_celler(self):
        self._rutenett = [[self.lag_celle() for j in range(len(self._rutenett[i]))] for i in range(len(self._rutenett))]

    def hent_celle(self, r, k):
        return None if r < 0 or r >= len(self._rutenett) or k < 0 or k >= len(self._rutenett[r]) else self._rutenett[r][k]
    
    def tegn_rutenett(self):
        for i in range(len(self._rutenett)):
            for j in range(len(self._rutenett[i])):
                print(self._rutenett[i][j].hent_status_tegn(), end="")
            print()

    def _sett_naboer(self, r, k):
        celle = self.hent_celle(r, k)
        for i in range(-1,2):
            for j in range(-1,2):
                if (i == 0 and j == 0): continue # Treffer selve cellen vi ønsker å sette opp naboene til.
                nr = r + i
                nk = k + j
                if nr >= 0 and nr < len(self._rutenett) and nk >= 0 and nk < len(self._rutenett[nr]):
                    celle.legg_til_nabo(self.hent_celle(nr, nk))

    def koble_celler(self):
        [[self._sett_naboer(i, j) for j in range(self._ant_kolonner)] for i in range(self._ant_rader)]

    def hent_alle_celler(self):
        return [self._rutenett[i][j] for j in range(self._ant_kolonner) for i in range(self._ant_rader)]
    
    def antall_levende(self):
        return sum(1 for x in self.hent_alle_celler() if x.er_levende())

File: GAME_OF_LIFE_in_Java/Rutenett.java
public class Rutenett { //Oppretter klassen med deklarasjonen av instansvariablene.
    Celle[][] rutene;
    public int antRader;
    public int antKolonner;

    public Rutenett(int r, int k) { //Konstruktøren til klassen som tar inn antall rader/kolonner i rutenettet som parametere.
        rutene = new Celle[r][k];
        antRader = r;
        antKolonner = k;
    }

    public void lagCelle(int r, int k) { //Lager en celle objekt i ønsket posisjon i rutenettet (parametere r,k(rad/kolonne)) med 1/3 sjanse å være levende.
        Celle celle = new Celle();
        if (Math.random() <= 0.3333) {
            celle.settLevende();
        }
        rutene[r][k] = celle;
    }       

    public void fyllMedTilfeldigeCeller() { //Fyller ut rutenettet med tilfeldige celler (levende/døde). 
        for (int i=0;i<rutene.length;i++) { //Dobbel for-løkke som itererer gjennom rader og kolonner.
            for (int j=0;j<rutene[i].length;j++) {
                lagCelle(i, j); //Bruker lagCelle-metoden for å opprette en tilfedig Celle og plassere det i en og en posisjon i rutenettet.
            }
        }
    }

    public Celle hentCelle(int r, int k) { //Returnerer et Celle-objekt ut ifra koordinatene (parametrene r og k (rad/kolonne)).
        if (r >= 0 && r < antRader && k >= 0 && k < antKolonner) { //Sørger for at indekser som er utenfor omfanget av rutenettet returnerer null. 
            return rutene[r][k];
        }
        return null;
    }

    public void tegnRutenett() { //Tegner rutenett i terminalen. 
        for(int i=0;i<3;i++) { //3 linjer mellomrom mellom hver iterasjon av rutenettet.
            System.out.println();
        }
        for (int i=0;i<rutene.length;i++) { 
            for (int j=0;j<rutene[i].length;j++) {
                System.out.print(rutene[i][j].hentStatusTegn()); //Printer cellene som ligger i hver rad etter hverandre.
            }
            System.out.println(); //Flytter kursoren til neste linje. 
        }
    }

    public void settNaboer(int r, int k) { //Setter opp naboer til hver celle i rutenettet. 
        Celle celle = hentCelle(r,k); //Henter opp cellen vi ønsker å legge naboer til  med metoden hentCelle.
        for (int i=-1;i<=1;i++) { //Dobbel for-løkke som itererer mellom verdiene -1 og 1 for hver rad/kolonne. Det er posisjonene over/under og til venstre/høyre fra 
            for (int j=-1;j<=1;j++) {                                                                           //Celle-objektet vi ønsker å legge naboer til.
                if (i == 0 && j == 0) { //Hvis variablene i og j begge blir like 0, så treffer løkken akkurat på Celle-objektet vi ønsker å legge naboer til.
                    continue;
                }
                int naboRad = r + i; //Oppretter indekser for nabo sine rad/kolonne posisjon. 
                int naboKol = k + j;

                if (naboRad >= 0 && naboRad < rutene.length && naboKol >= 0 && naboKol < rutene[naboRad].length) { //Sjekker om at indeksene til naboen er gyldige. 
                    Celle naboCelle = hentCelle(naboRad, naboKol); //Hvis indeksene er gyldige hentes det naboen til cellen med hentCelle metoden.
                    celle.leggTilNabo(naboCelle); //Naboen legges til ved bruk av leggTilNabo metoden fra Celle-klassen.
                }
            }
        }
    }

    public void kobleAlleCeller() { //Metoden itererer gjennom hver posisjon i rutenettet og setter opp naboer for hver Celle-objekt i rutenettet.
        for (int i=0;i<rutene.length;i++) {
            for (int j=0;j<rutene[i].length;j++) {
                settNaboer(i, j);
            }
        }
    }

    public int antallLevende() { //Metoden som itererer gjennom rutenettet og returnerer totalt antall levende celler i rutenettet.
        int teller = 0;
        for (int i=0;i<rutene.length;i++) {
            for (int j=0;j<rutene[i].length;j++) {
                if (rutene[i][j].erLevende()) {
                    teller++;
                }
            }
        } return teller;
    }

}